src/utils.ts

```diff
@@ -148,10 +168,76 @@ export async function fetchFlutterServerPort({
          return forwardedPort!;
       } catch (e) {
          if (portReleaseCallback) {
-            await portReleaseCallback(udid, systemPort!);
+            await portReleaseCallback(udid, forwardedPort!);
+         }
+         if (portForwardCallback) {
+            await portForwardCallback(udid, systemPort!, devicePort);
+         }
+         try {
+            log.info(
+               `Checking if flutter server is running on port ${devicePort}`,
+            );
+            await waitForFlutterServer(
+               forwardedPort!,
+               packageName,
+               flutterCaps,
+            );
+            log.info(
+               `Flutter server is successfully running on port ${devicePort}`,
+            );
+            return forwardedPort!;
+         } catch (e) {
+            if (portReleaseCallback) {
+               await portReleaseCallback(udid, systemPort!);
+            }
          }
+         devicePort++;
       }
-      devicePort++;
    }
    return null;
 }
+
+export function isW3cCaps(caps: any) {
+   if (!_.isPlainObject(caps)) {
+      return false;
+   }
+
+   const isFirstMatchValid = () =>
+      _.isArray(caps.firstMatch) &&
+      !_.isEmpty(caps.firstMatch) &&
+      _.every(caps.firstMatch, _.isPlainObject);
+   const isAlwaysMatchValid = () => _.isPlainObject(caps.alwaysMatch);
+   if (_.has(caps, 'firstMatch') && _.has(caps, 'alwaysMatch')) {
+      return isFirstMatchValid() && isAlwaysMatchValid();
+   }
+   if (_.has(caps, 'firstMatch')) {
+      return isFirstMatchValid();
+   }
+   if (_.has(caps, 'alwaysMatch')) {
+      return isAlwaysMatchValid();
+   }
+   return false;
+}
+
+export function attachAppLaunchArguments(parsedCaps: any, ...caps: any) {
+   const capsToUpdate = [...caps].find(isW3cCaps);
+   const platformName: string | undefined = parsedCaps['platformName'];
+   const systemPort: string | undefined = parsedCaps['flutterSystemPort'];
+   console.log(
+      'Caps to update:',
+      systemPort,
+      capsToUpdate,
+      `${capsToUpdate.alwaysMatch['appium:flutterSystemPort']}`,
+   );
+   if (platformName && systemPort && platformName.toLowerCase() == 'ios') {
+      const args = [
+         `--port=${capsToUpdate.alwaysMatch['appium:flutterSystemPort']}`,
+      ];
+      log.info(
+         `iOS platform detected and flutterSystemPort capability is present. So attaching processArguments: ${JSON.stringify(args)}`,
+      );
+      capsToUpdate.alwaysMatch['appium:processArguments'] = {
```

Comment:
- it would be safer to update the value in both firstMatch and alwaysMatch objects
- we must not rewrite it as it is going to corrupt the customer input if they provide any value in caps. Instead we must check if the value already exists and add the desired arg there.

Reply:
`it would be safer to update the value in both firstMatch and alwaysMatch objects` @mykola-mokhnach Wouldn't appium throw error saying the value is present in both alwaysMatch and firstMatch? I have seen these kind of errors not sure if it holds good for this case.

Reply:
`we must not rewrite it as it is going to corrupt the customer input if they provide any value in caps. Instead we must check if the value already exists and add the desired arg there` Will address as part of issue #34

Reply:
> Wouldn't appium throw error saying the value is present in both alwaysMatch and firstMatch? I have seen these kind of errors not sure if it holds good for this case.

It will. That is why the following algorithm should be applied:
- check alwaysMatch. Change the value if present there.
- check firstMatch. Change the value if present in any of the firstMatch array items
- If the value is not present in both objects then add it to alwaysMatch